File: app/src/main/java/com/jahir/pia_appsmov/IS_R/InicioSesion.java
package com.jahir.pia_appsmov.IS_R;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.firestore.auth.User;
import com.jahir.pia_appsmov.Menu.Menu;
import com.jahir.pia_appsmov.Model.Usuario;
import com.jahir.pia_appsmov.R;

public class InicioSesion extends AppCompatActivity {

    Button ISBtn, RBtn;
    EditText Correo, Contrasena;
    FirebaseAuth mAuth;
    ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_inicio_sesion);

        Correo = findViewById(R.id.Correo);
        Contrasena = findViewById(R.id.Contrasena);
        ISBtn = findViewById(R.id.ISBtn);
        RBtn = findViewById(R.id.RBtn);
        mAuth = FirebaseAuth.getInstance();

        ISBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String correo = Correo.getText().toString().trim();
                String password = Contrasena.getText().toString().trim();

                if(correo.isEmpty() && password.isEmpty()){
                    Toast.makeText(InicioSesion.this, "Complete los campos correspondientes", Toast.LENGTH_SHORT).show();
                }else{
                    InicioSesionU(correo, password);
                    progressDialog = new ProgressDialog(InicioSesion.this);
                    progressDialog.show();
                    progressDialog.setContentView(R.layout.pantalla_de_carga);
                    progressDialog.getWindow().setBackgroundDrawableResource(
                            android.R.color.transparent
                    );
                }
            }
        });

        RBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(InicioSesion.this, Registrarse.class));
            }
        });

    }

    private void InicioSesionU(String correo, String password) {
        mAuth.signInWithEmailAndPassword(correo, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if(task.isSuccessful()){
                    finish();
                    startActivity(new Intent(InicioSesion.this, Menu.class));
                    progressDialog.dismiss();
                    Toast.makeText(InicioSesion.this, "Bienvenido", Toast.LENGTH_SHORT).show();
                }else{
                    Toast.makeText(InicioSesion.this, "Error", Toast.LENGTH_SHORT).show();
                }

            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                Toast.makeText(InicioSesion.this, "Error al iniciar sesión", Toast.LENGTH_SHORT).show();
            }
        });

    }

    /*@Override
    protected void onStart() {
        super.onStart();
        FirebaseUser usuario = mAuth.getCurrentUser();
        if(usuario != null){
            startActivity(new Intent(InicioSesion.this, Registrarse.class));
            finish();
        }
    }*/

}

File: app/src/main/java/com/jahir/pia_appsmov/usescases/iniciosesion/InicioSesion.kt
package com.jahir.pia_appsmov.usescases.iniciosesion

import android.app.ProgressDialog
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.auth.FirebaseAuth
import com.jahir.pia_appsmov.usescases.menu.Menu
import com.jahir.pia_appsmov.R
import com.jahir.pia_appsmov.usescases.registro.Registrarse
import com.jahir.pia_appsmov.usescases.registro.RetrofitClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class InicioSesion : AppCompatActivity() {

    lateinit var ISBtn: Button
    lateinit var RBtn: Button
    lateinit var User: TextInputEditText
    lateinit var Correo: TextInputEditText
    lateinit var Contrasena: TextInputEditText
    lateinit var progressDialog: ProgressDialog
    lateinit var preferences: SharedPreferences
    var INS = NomU("","","")
    var siempre = true

    lateinit var mAuth : FirebaseAuth

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_inicio_sesion)

        Correo = findViewById(R.id.Correo)
        User = findViewById(R.id.User)
        Contrasena = findViewById(R.id.Contrasena)
        ISBtn = findViewById(R.id.ISBtn)
        RBtn = findViewById(R.id.RBtn)
        mAuth = FirebaseAuth.getInstance()

        preferences = getSharedPreferences("Preferences", MODE_PRIVATE)

        ISBtn.setOnClickListener(View.OnClickListener {
            val correo = Correo.getText().toString().trim { it <= ' ' }
            val password = Contrasena.getText().toString().trim { it <= ' ' }
            val user = User.getText().toString().trim { it <= ' ' }
            if (correo.isEmpty() && password.isEmpty() && user.isEmpty()) {
                Toast.makeText(
                    this@InicioSesion,
                    "Complete los campos correspondientes",
                    Toast.LENGTH_SHORT
                ).show()
            } else {
                InicioSesionU(user, correo, password)
                progressDialog = ProgressDialog(this@InicioSesion)
                progressDialog!!.show()
                progressDialog!!.setContentView(R.layout.pantalla_de_carga)
                progressDialog!!.window!!.setBackgroundDrawableResource(
                    android.R.color.transparent
                )
            }
        })
        RBtn.setOnClickListener(View.OnClickListener {
            startActivity(
                Intent(
                    this@InicioSesion,
                    Registrarse::class.java
                )
            )
        })
    }

    private fun InicioSesionFirebase(user: String, correo: String, password: String) {
        mAuth!!.signInWithEmailAndPassword(correo, password).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                Toast.makeText(this@InicioSesion, "Cuenta existente",Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun InicioSesionU(user: String, correo: String, password: String) {
        this.INS.user = user
        this.INS.email = correo
        this.INS.passW = password

        if(siempre) {
            CoroutineScope(Dispatchers.IO).launch {
                val call = RetrofitClient.webService.INUser(INS)
                runOnUiThread {
                    if (call.body().toString().equals("SI")) {

                        val editor = preferences!!.edit()
                        editor.putString("user", user)
                        editor.putString("correo", correo)
                        editor.putString("mipassword", password)
                        editor.commit()

                        InicioSesionFirebase(user, correo, password)

                        Toast.makeText(this@InicioSesion, "Bienvenido", Toast.LENGTH_SHORT).show()
                        limpiarObjeto()
                        val intent = Intent(this@InicioSesion, Menu::class.java)
                        startActivity(intent)
                        limpiarcampos()
                        finish()
                        progressDialog!!.dismiss()
                    } else {
                        Toast.makeText(
                            this@InicioSesion,
                            call.body().toString(),
                            Toast.LENGTH_SHORT
                        ).show()
                        progressDialog!!.dismiss()
                    }
                }
            }
        }
    }



    private fun limpiarcampos() {
        Correo.setText("")
        User.setText("")
        Contrasena.setText("")
    }

    private fun limpiarObjeto() {
        this.INS.user = ""
        this.INS.email = ""
        this.INS.passW = ""
    }


}